x-pack/plugins/alerting/README.md

```diff
@@ -0,0 +1,227 @@
+
+## Alert instance factory
+
+**alertInstanceFactory(id)**
+
+One service passed in to alert types is an alert instance factory. This factory creates instances of alerts and must be used in order to fire actions. These instances support state persisting between alert type execution but will clear out once the alert instance stops firing.
+
+This factory returns an instance of `AlertInstance`. The alert instance class has the following methods, note that we have removed the methods that you shouldn't touch.
+
+|Method|Description|
+|---|---|
+|getState()|Get the current state of the alert instance.|
+|fire(actionGroup, context)|Called to fire actions. The actionGroup relates to the group of alert `actions` to fire and the context will be used for templating purposes. This should only be called once per alert instance.|
+|replaceState(state)|Used to replace the current state of the alert instance. This doesn't work like react, the entire state must be provided. Use this feature as you see fit. The state that is set will persist between alert type executions whenever you re-create an alert instance with the same id. The instance state will be erased when fire isn't called during an execution.|
+
+## Templating actions
+
+There needs to be a way to map alert context into action parameters. For this, we started off by adding template support. Any string within the `params` of an alert saved object's `actions` will be processed as a template and can inject context or state values. 
+
+When an alert instance fires, the first argument is the `group` of actions to fire and the second is the context the alert exposes to templates. We iterate through each action params attributes recursively and render templates if they are a string. Templates have access to the `context` (provided by second argument of `.fire(...)` on an alert instance) and the alert instance's `state` (provided by the most recent `replaceState` call on an alert instance).
+
+### Examples
+
+The following code would be within an alert type. As you can see `cpuUsage ` will replace the state of the alert instance and `server` is the context for the alert instance to fire. The difference between the two is `cpuUsage ` will be accessible at the next execution.
+
+```
+alertInstanceFactory('server_1')
+  .replaceState({
+    cpuUsage: 80,
+  })
+  .fire('default', {
+    server: 'server_1',
+  });
+```
+
+Below is an example of an alert that takes advantage of templating:
+
+```
+{
+  ...
+  actions: [
+    {
+      "group": "default",
```

Comment:
so thinking about this more: group names are configured on alert instances, but the alert type code itself needs to know names for when it fires.  Don't see how this will work in practice unless we only have predefined names that you can use.

Reply:
Yes I think when we come up with throttling we'll make the alert type the source of truth for these groups and define a priority or something so we know when to stop throttling when alert status gets worse.